### PR TITLE
S3 object store

### PR DESCRIPTION
**Pull request by mgasner**

### CODE REVIEW

File: python_modules/dagster/dagster/core/object_store.py [hunk 1]
```diff
@@ -1,6 +1,14 @@
 import os
 import pickle
 
+from io import BytesIO
+
+try:
```
Comment (schrockn) [inline python_modules/dagster/dagster/core/object_store.py]: what would you think about doing this include on-demand within S3ObjectStore methods and then failing in a loud way?

File: python_modules/dagster/dagster/core/object_store.py [hunk 2]
```diff
@@ -45,6 +55,72 @@ def get_object(self, context, runtime_type, paths):  # pylint: disable=unused-ar
...
+        return pickle.loads(self.s3.get_object(Bucket=self.bucket, Key=key)['Body'].read())
+
+    def has_object(self, context, paths):  # pylint: disable=unused-argument
+        key = self._key_for_paths(paths)
+
+        try:
+            self.s3.head_object(Bucket=self.bucket, Key=key)
+            return True
+        except botocore.exceptions.ClientError as exc:
+            if 'Error' in exc.response:
```
Comment (schrockn) [inline python_modules/dagster/dagster/core/object_store.py]: could do: 
```
exec.response.get('Error', {}).get('Code') == '404'
```

File: python_modules/dagster/dagster/core/object_store.py [hunk 3]
```diff
@@ -45,6 +55,72 @@ def get_object(self, context, runtime_type, paths):  # pylint: disable=unused-ar
...
+                if 'Code' in exc.response['Error']:
```
Comment (schrockn) [inline python_modules/dagster/dagster/core/object_store.py]: Also throw if it is a different error?

File: python_modules/dagster/dagster/core/object_store.py [hunk 4]
```diff
@@ -45,6 +55,72 @@ def get_object(self, context, runtime_type, paths):  # pylint: disable=unused-ar
...
+                    if exc.response['Error']['Code'] == '404':
+                        return False
+
+    def rm_object(self, context, paths):
```
Comment (schrockn) [inline python_modules/dagster/dagster/core/object_store.py]: add this to the base class?

File: python_modules/dagster/dagster_tests/core_tests/test_object_store.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        try:
+            shutil.rmtree(object_store.root)
+        except seven.FileNotFoundError:
+            pass
+
+
+def test_s3_object_store():
+    run_id = str(uuid.uuid4())
+
+    object_store = S3ObjectStore(run_id=run_id, s3_bucket='dagster-airflow-scratch')
```
Comment (schrockn) [inline python_modules/dagster/dagster_tests/core_tests/test_object_store.py]: hmmm. this breaks the "laptop" use case. i think we should mark this as an circle-test only

Comment (schrockn) [reply python_modules/dagster/dagster_tests/core_tests/test_object_store.py]: how do creds work here?

File: python_modules/dagster/tox.ini [hunk 6]
```diff
@@ -2,9 +2,9 @@
 envlist = py37,py36,py35,py27
 
 [testenv]
-passenv = CIRCLECI CIRCLE_* CI_PULL_REQUEST COVERALLS_REPO_TOKEN
+passenv = CIRCLECI CIRCLE_* CI_PULL_REQUEST COVERALLS_REPO_TOKEN AWS_SECRET_ACCESS_KEY AWS_ACCESS_KEY_ID
```
Comment (schrockn) [inline python_modules/dagster/tox.ini]: ah this is how creds work

File: python_modules/dagster/dagster/core/object_store.py (hunk 1 above)
Comment (mgasner) [reply python_modules/dagster/dagster/core/object_store.py]: happy to do that

File: python_modules/dagster/dagster/core/object_store.py (hunk 2 above)
Comment (mgasner) [reply python_modules/dagster/dagster/core/object_store.py]: yep

File: python_modules/dagster/dagster/core/object_store.py (hunk 3 above)
Comment (mgasner) [reply python_modules/dagster/dagster/core/object_store.py]: yep

File: python_modules/dagster/dagster/core/object_store.py (hunk 4 above)
Comment (mgasner) [reply python_modules/dagster/dagster/core/object_store.py]: yep

File: python_modules/dagster/dagster_tests/core_tests/test_object_store.py (hunk 5 above)
Comment (mgasner) [reply python_modules/dagster/dagster_tests/core_tests/test_object_store.py]: ordinary boto chain, expecting `AWS_...` env variables to be set. we could additionally make S3ObjectStore directly initable with creds

Comment (schrockn) [reply python_modules/dagster/dagster_tests/core_tests/test_object_store.py]: got it. no need to add aws creds until we need it

**Pull request merged**